osu.Game/IO/OsuStorage.cs

```diff
@@ -9,17 +13,111 @@ namespace osu.Game.IO
 {
     public class OsuStorage : WrappedStorage
     {
+        private readonly GameHost host;
+        private readonly StorageConfigManager storageConfig;
+
+        internal static readonly string[] IGNORE_DIRECTORIES = { "cache" };
+
+        internal static readonly string[] IGNORE_FILES =
+        {
+            "framework.ini",
+            "storage.ini"
+        };
+
         public OsuStorage(GameHost host)
             : base(host.Storage, string.Empty)
         {
-            var storageConfig = new StorageConfigManager(host.Storage);
+            this.host = host;
+
+            storageConfig = new StorageConfigManager(host.Storage);
 
             var customStoragePath = storageConfig.Get<string>(StorageConfig.FullPath);
 
             if (!string.IsNullOrEmpty(customStoragePath))
-            {
                 ChangeTargetStorage(host.GetStorage(customStoragePath));
-                Logger.Storage = UnderlyingStorage.GetStorageForDirectory("logs");
+        }
+
+        protected override void ChangeTargetStorage(Storage newStorage)
+        {
+            base.ChangeTargetStorage(newStorage);
+            Logger.Storage = UnderlyingStorage.GetStorageForDirectory("logs");
+        }
+
+        public void Migrate(string newLocation)
+        {
+            var source = new DirectoryInfo(GetFullPath("."));
+            var destination = new DirectoryInfo(newLocation);
+
+            // ensure the new location has no files present, else hard abort
+            if (destination.Exists)
+            {
+                if (destination.GetFiles().Length > 0)
+                    throw new InvalidOperationException("Migration destination already has files present");
+
+                deleteRecursive(destination);
```

Comment:
Is this hoping to delete the destination directory itself for the `copyRecursive` call below? Because this doesn't seem like it would do that (only delete all files/directories within the destination).

Reply:
the directory itself is fine to continue existing in this case, i think.